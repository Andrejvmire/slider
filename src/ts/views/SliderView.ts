import AbstractViewPublisher from "../abstract/AbstractViewPublisher";
import PointView from "./PointView";
import RulerView from "./RulerView";
import TooltipView from "./TooltipView";
import RangerView from "./RangerView";

export default class SliderView extends AbstractViewPublisher implements IViewPublisher, IViewSubscriber {
    private value: [number] | [number, number];
    private _points: (IViewPublisher & IPoint)[] = [];
    private _tooltips: IView[] = [];
    private _ruler: IView;
    private _ranger: IView & IViewSubscriber;
    private _side: "left" | "top";
    private readonly $_container: JQuery;
    private static className: string = 'slider slider__container';

    constructor(private options: ViewOptionsType, private parent: JQuery) {
        super();
        this.value = (typeof options.points === "number") ? [options.points] : options.points;
        this.$_container = parent
            .addClass(SliderView.className);
        this.pointsInit();
        this.rulerInit();
        this.tooltipInit();
        this.rangerInit();
        this.render();
    }

    private static appendView(view: IView, parent: IView): JQuery {
        if (typeof view !== "undefined") {
            view.$instance
                .appendTo(parent.$instance);
        }
        return parent.$instance;
    }

    private render(): void {
        SliderView.appendView(this._ranger, this._ruler);
        this._ruler.$instance
            .append(
                this._points.map(
                    (point, index) => {
                        return SliderView.appendView(this._tooltips[index], point)
                    }
                )
            )
            .appendTo(this.$_container);
        this.events();
    }

    private rangerInit(): void {
        if (this._points.length === 2) {
            let value = <[number, number]>this._points
                .map(
                    point => point.state
                )
            this._ranger = new RangerView(value, this._side);
        }
    }

    private tooltipInit(): void {
        let {tooltip} = this.options;
        if (tooltip) {
            this._tooltips = this.value.map(
                point => new TooltipView(point, this._side)
            )
        }
    }

    private rulerInit(): void {
        this._ruler = new RulerView(this.options.ruler);
    }

    private pointsInit(): void {
        let {orientation, ruler} = this.options;
        this._side =
            ((orientation === 'horizontal') || (typeof orientation === "undefined"))
                ?
                "left"
                :
                "top";
        this._points = this.value
            .sort()
            .map(
                point => new PointView(SliderView.pointInPercents(point, ruler), this._side, this)
            )
    }

    private static pointInPercents(part: number, from: [number, number]): number {
        return (part - from[0]) * 100 / from.reduce((p, c) => c - p);
    }

    private percentsInPoint(event: JQuery.TriggeredEvent): number {
        let offset = this.$_container.offset(),
            relativeX = (event.pageX || 0) - (offset?.left || 0),
            relativeY = (event.pageY || 0) - (offset?.top || 0),
            width = this.$_container.width() || 1,
            height = this.$_container.height() || 1,
            percent = {
                left: (relativeX * 100) / width,
                top: (relativeY * 100) / height
            };
            return percent[this._side];
    }

    update(value: number): void {
        let [from, to] = this.options.ruler;
        this.state = <[number] | [number, number]>this._points.map(
            point => {
                let number: number;
                if (typeof point.state !== "number") {
                    number = point.state[0];
                } else {
                    number = point.state;
                }
                return Math.floor(number * (to - from) / 100) + from;
            }
        )
        this.notify();
    }

    private events(): void {
        this._points
            .map(
                point => {
                    point.$instance
                        .on('mousedown.slider__point', () => {
                            $(document).one('mousedown.slider__point', () => false);
                            $(document).one('mouseup.slider__point', () => {
                                $(document).off('mousemove.slider__point');
                            })
                            $(document).on('mousemove.slider__point', mouseMoveEvent => {
                                point.moveTo(this.percentsInPoint(mouseMoveEvent));
                                point.notify();
                            })
                        })
                }
            )
        this._ruler.$instance
            .on('click.slider__ruler', clickEvent => {
                let newPoint = this.percentsInPoint(clickEvent);
                this._points
                    .reduce(
                        ((previousValue, currentValue) =>
                            (Math.abs(previousValue.state - newPoint) <= Math.abs(currentValue.state - newPoint))
                                ? previousValue
                                : currentValue)
                    )
                    .moveTo(newPoint);
            })
    }

}