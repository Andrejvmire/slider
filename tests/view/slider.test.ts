import SliderView from "../../src/ts/views/SliderView";
import $ from 'jquery';

describe('Тестируем SliderView', function () {
    let parent = $(document.createElement('div'));
    let slider: IViewPublisher & IViewSubscriber & ISlider;
    describe('Создаем класс', function () {
        beforeEach(() => {
            slider = new SliderView({
                points: [34, 45],
                tooltip: true,
                orientation: "vertical",
                ruler: [0, 120],
            }, parent);
        });
        it('должен веруть объект не undefined', function () {
            expect(slider)
                .not.toBeUndefined();
        });
        it('должен вернуть $instance типа JQuery', function () {
            expect(slider.$instance)
                .toBeInstanceOf($);
        });
        it("должен соответствовать снимку", function () {
            expect(parent)
                .toMatchSnapshot();
        })
    });
    describe('Меняем состояние', function () {
        beforeEach(() => {
            slider = new SliderView({
                points: [-3, 77],
                ruler: [-23, 199],
                orientation: "vertical"
            }, parent)
        });
        it('Переместит точку', function () {
            expect(slider.move(-20, -3).state)
                .toStrictEqual([-20, 77]);
        });
    });
});