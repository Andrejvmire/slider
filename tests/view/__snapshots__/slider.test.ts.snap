// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестируем SliderView Создаем класс дложен соответствовать слепку 1`] = `
jQuery {
  "0": <div
    class="slider slider__container slider__container_vertical"
  >
    <div
      class="slider slider__ruler slider__ruler_vertical"
    >
      <div
        class="slider slider__ranger"
        style="top: 28.333333333333332%; height: 9.166666666666668%;"
      />
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 28.333333333333332%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          34
        </span>
      </span>
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 37.5%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          45
        </span>
      </span>
    </div>
    <div
      class="slider slider__ruler slider__ruler_vertical"
    >
      <div
        class="slider slider__ranger"
        style="top: 28.333333333333332%; height: 9.166666666666668%;"
      />
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 28.333333333333332%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          34
        </span>
      </span>
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 37.5%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          45
        </span>
      </span>
    </div>
    <div
      class="slider slider__ruler slider__ruler_vertical"
    >
      <div
        class="slider slider__ranger"
        style="top: 28.333333333333332%; height: 9.166666666666668%;"
      />
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 28.333333333333332%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          34
        </span>
      </span>
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 37.5%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          45
        </span>
      </span>
    </div>
  </div>,
  "length": 1,
}
`;

exports[`Тестируем SliderView Создаем класс должен соответствовать слепку 1`] = `
jQuery {
  "0": <div
    class="slider slider__container slider__container_vertical"
  >
    <div
      class="slider slider__ruler slider__ruler_vertical"
    >
      <div
        class="slider slider__ranger"
        style="top: 28.333333333333332%; height: 9.166666666666668%;"
      />
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 28.333333333333332%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          34
        </span>
      </span>
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 37.5%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          45
        </span>
      </span>
    </div>
    <div
      class="slider slider__ruler slider__ruler_vertical"
    >
      <div
        class="slider slider__ranger"
        style="top: 28.333333333333332%; height: 9.166666666666668%;"
      />
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 28.333333333333332%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          34
        </span>
      </span>
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 37.5%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          45
        </span>
      </span>
    </div>
    <div
      class="slider slider__ruler slider__ruler_vertical"
    >
      <div
        class="slider slider__ranger"
        style="top: 28.333333333333332%; height: 9.166666666666668%;"
      />
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 28.333333333333332%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          34
        </span>
      </span>
      <span
        class="slider slider__point slider__point_vertical"
        style="top: 37.5%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_vertical"
        >
          45
        </span>
      </span>
    </div>
  </div>,
  "length": 1,
}
`;
